metaflow/plugins/kfp/kfp_decorator.py

```diff
@@ -0,0 +1,89 @@
+import json
+import os
+import tarfile
+import tempfile
+from urllib.parse import urlparse
+
+from metaflow import util, current
+from metaflow.datastore import MetaflowDataStore
+from metaflow.datastore.util.s3util import get_s3_client
+from metaflow.decorators import StepDecorator
+from metaflow.metadata import MetaDatum
+from metaflow.metaflow_config import DATASTORE_LOCAL_DIR
+
+
+class KfpInternalDecorator(StepDecorator):
+    name = "kfp_internal"
+
+    def task_pre_step(
+        self,
+        step_name,
+        ds,
+        metadata,
+        run_id,
+        task_id,
+        flow,
+        graph,
+        retry_count,
+        max_retries,
+    ):
+        """
+        Analogous to step_functions_decorator.py
+        """
+        # TODO: any other KFP environment variables to get and register to Metadata service?
+        meta = {"kfp-execution": os.environ["METAFLOW_RUN_ID"]}
+        entries = [MetaDatum(field=k, value=v, type=k) for k, v in meta.items()]
+        # Register book-keeping metadata for debugging.
+        metadata.register_metadata(run_id, step_name, task_id, entries)
+
+        if metadata.TYPE == "local":
+            self.ds_root = ds.root
+        else:
+            self.ds_root = None
+
+    def task_finished(
+        self, step_name, flow, graph, is_task_ok, retry_count, max_user_code_retries
+    ):
+        """
+        Analogous to step_functions_decorator.py
+        """
+        if not is_task_ok:
+            # The task finished with an exception - execution won't
+            # continue so no need to do anything here.
```

Comment:
Can we log the error, or the raise the exception here? Not sure if it is raised elsewhere, but good to let the end user know what went wrong.

Reply:
`task_finished` runs after the task has finished successfully or not, and doesn't have access to the error.
The role of running the task is in `task.py` `run_step()` https://github.com/zillow/metaflow/blob/a63e24e75f3e206d948a15158bff311c58767d1e/metaflow/task.py#L377-L377, and it logs the exception in a try/catch.

Reply:
Makes sense, so this logs the exception internally. As long as the user sees it, then we're good.